src/Sarif/Query/Evaluators/PropertyBagPropertyEvaluator.cs

```diff
@@ -0,0 +1,121 @@
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        SdkResources.ErrorInvalidQueryPropertyPrefix,
+                        "'" + string.Join("', '", ResultPropertyPrefix, RulePropertyPrefix)) + "'",
+                    nameof(term));
+            }
+
+            _evaluator = CreateEvaluator(term);
+        }
+
+        // Create an appropriate evaluator object for the specified term. For operators that apply
+        // only to strings (":" (contains), "|>" (startswith), or ">|" (endswith), always create
+        // a string evaluator.  All other operators (==, !=, >, <, etc.) can apply to both strings
+        // and numbers. If the value being compared parses as a number, assume that a numeric
+        // comparison was intended, and create a numeric evaluator. Otherwise, create a string evaluator.
+        // This could cause problems if the comparand is string that happens to look like a number.
+        private IExpressionEvaluator<Result> CreateEvaluator(TermExpression term) =>
+           IsStringComparison(term)
+            ? new StringEvaluator<Result>(GetProperty<string>, term, StringComparison.OrdinalIgnoreCase) as IExpressionEvaluator<Result>
+            : new DoubleEvaluator<Result>(GetProperty<double>, term);
+
+        private static readonly ReadOnlyCollection<CompareOperator> s_stringSpecificOperators =
+            new ReadOnlyCollection<CompareOperator>(
+                new CompareOperator[]
+                {
+                    CompareOperator.Contains,
+                    CompareOperator.EndsWith,
+                    CompareOperator.StartsWith
+                });
+
+        private bool IsStringComparison(TermExpression term)
+            => s_stringSpecificOperators.Contains(term.Operator) || !double.TryParse(term.Value, out _);
+
+        private T GetProperty<T>(Result result)
+        {
+            PropertyBagHolder holder = GetPropertyBagHolder(result);
+            return GetPropertyFromHolder<T>(holder);
+        }
+
+        private PropertyBagHolder GetPropertyBagHolder(Result result) =>
+            _propertyBelongsToRule
+            ? result.GetRule() as PropertyBagHolder
+            : result;
+
+        private T GetPropertyFromHolder<T>(PropertyBagHolder holder)
+        {
+            T value = default;
+
+            // We can't just call holder.TryGetProperty because we want a case-insensitive
+            // match.
+            List<string> propertyNames = holder.PropertyNames
+                .Where(key => key.Equals(_propertyName, StringComparison.OrdinalIgnoreCase))
+                .ToList();
+
+            if (propertyNames.Any())
+            {
+                try
+                {
+                    // If the property bag contains two properties whose names differ only
+                    // by case, we'll take the first one.
```

Comment:
>we'll take the first on [](start = 32, length = 23)

why have we implemented a case-insensitive property name search? it seems like this is simply adding more complexity for us (and introducing surprising results such as returning an arbitrary match).

Reply:
Agreed, fixing.

---
In reply to: [496995760](https://github.com/microsoft/sarif-sdk/pull/2065#discussion_r496995760) [](ancestors = 496995760)